src/cargo/sources/git/utils.rs

```diff
@@ -694,12 +695,38 @@ pub fn with_fetch_options(
     let mut progress = Progress::new("Fetch", config);
     network::with_retry(config, || {
         with_authentication(url, git_config, |f| {
+            let mut last_recv = 0; // in Byte
+            let mut last_update = Instant::now();
             let mut rcb = git2::RemoteCallbacks::new();
             rcb.credentials(f);
-
             rcb.transfer_progress(|stats| {
+                let indexed_deltas = stats.indexed_deltas();
+                let msg = if indexed_deltas > 0 {
+                    // Resolving deltas.
+                    format!(
+                        ", ({}/{}) resolving deltas",
+                        indexed_deltas,
+                        stats.total_deltas()
+                    )
+                } else {
+                    // Receiving objects.
+                    let duration = last_update.elapsed();
+                    let recv = stats.received_bytes();
+                    let rate = (recv - last_recv) as f32 / duration.as_secs_f32();
+                    if duration > Duration::from_secs(3) {
```

Comment:
I'm a bit worried about a calcuation like this jumping around by accident. For example if the next amount of data received after this tick just happened to be overly large or overly small it could cause what appeared to be large fluctuations in the reported transfer rate.

Thinking more on this it's not actually something we print for HTTP downloads either (data receiving rate). While an abstraction could be used here to make sure that the reported rate is a continuous calculation of whatever we received in the last N seconds, I think it may make more sense to print this out either additionally for HTTP download (and share the same abstraction for printing between the two) or otherwise just print the total amount downloaded here without the rate (although having a rate does seem quite nice).

Reply:
Sorry, I didn't get all your points. To avoid large fluctuations maybe we can adjust N to 5 seconds. That's how curl's progress meter does. 

- https://everything.curl.dev/cmdline/progressmeter
- https://github.com/curl/curl/blob/f014eec/src/tool_progress.c#L243-L269

For the latter, `stats.received_bytes()` may not be the actual HTTP received bytes but it is network packet size [according to libgit2 implementation](https://github.com/libgit2/libgit2/blob/5083614/src/transports/smart_protocol.c#L498-L499), so I think it's ok to be represented as data receiving rate. Did you mean that we should add more text to indicate what it does? 

I think I am a bit confused now.... Could provide the preferred UI in your mind?

Reply:
Sorry I'll try to be more clear. With the logic you have implemented the transfer rate for each three-second window is a discrete nonoverlapping three second window with all other three second windows. I think that this could cause large fluctuations in the output of the transfer rate. For example if, for whatever reason, your network slowed to a crawl at the beginning of each 3-second window and then greatly increased in speed by the end we'd constantly be showing a fluctuating transfer speed here.

I think instead this should do what curl is doing, which is to have a table of historical measurments of speeds. There's a running counter which includes only the last three seconds of measurements, and the current time of measurment is recorded for new measurements and used to discard old measurements. This means that we'd be displaying a three-second average of the transfer speed starting at any given point in time going back three seconds instead of having discrete 3-second windows.

An abstraction like this should most likely be separated from this definition location itself to be somewhere in cargo's `utils` module, for example.

Reply:
To be clear, though, I think the UI is fine it's just the measuring of the number to get displayed I think needs to be tweaked.

Reply:
Oh, I finally got it. Thank for your detailed explanation! I'll get it done soon.

Reply:
Just add a `MetricsCounter` to do what curl does!